lib/shoryuken/middleware/server/exponential_backoff_retry.rb

```diff
@@ -0,0 +1,54 @@
+module Shoryuken
+  module Middleware
+    module Server
+      class ExponentialBackoffRetry
+        include Util
+
+        def call(worker, queue, sqs_msg, body)
+          started_at = Time.now
+          yield
+        rescue Exception => e
+          retry_intervals = Array(worker.class.get_shoryuken_options['retry_intervals'])
+           
+          if retry_intervals.empty? || ! handle_failure(sqs_msg, started_at, retry_intervals)
+            # Re-raise the exception if the job is not going to be retried.
+            # This allows custom middleware (like exception notifiers) to be aware of the unhandled failure.
+            raise e
+          end
+        end
+        
+      private
+      
+        def handle_failure(sqs_msg, started_at, retry_intervals)
+          attempts = sqs_msg.attributes['ApproximateReceiveCount']
```

Comment:
`sqs_msg.attributes` is returning `nil` for me, we probably need to 

https://github.com/phstc/shoryuken/blob/master/lib/shoryuken/fetcher.rb#L18

```ruby
options[:attribute_names] = %w[ApproximateFirstReceiveTimestamp ApproximateReceiveCount]

# OR %w[All]
```

http://docs.aws.amazon.com/sdkforruby/api/Aws/SQS/Client.html#receive_message-instance_method

> :attribute_names => Array<String>
> A list of attributes that need to be returned along with each message.
> 
> The following lists the names and descriptions of the attributes that can be returned:
> 
> All - returns all values.
> ApproximateFirstReceiveTimestamp - returns the time when the message was first received from the queue (epoch time in milliseconds).
> ApproximateReceiveCount - returns the number of times a message has been received from the queue but not deleted.
> SenderId - returns the AWS account number (or the IP address, if anonymous access is allowed) of the sender.
> SentTimestamp - returns the time when the message was sent to the queue (epoch time in milliseconds).


```ruby
resp = sqs.receive_message(
  # required
  queue_url: "String",
  attribute_names: ["Policy|VisibilityTimeout|MaximumMessageSize|MessageRetentionPeriod|ApproximateNumberOfMessages|ApproximateNumberOfMessagesNotVisible|CreatedTimestamp|LastModifiedTimestamp|QueueArn|ApproximateNumberOfMessagesDelayed|DelaySeconds|ReceiveMessageWaitTimeSeconds|RedrivePolicy", '...'],
  message_attribute_names: ["MessageAttributeName", '...'],
  max_number_of_messages: 1,
  visibility_timeout: 1,
  wait_time_seconds: 1,
)
```

wdyt?